src/charm.py

```diff
@@ -5,19 +5,176 @@
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
+BASE_CONFIG_PATH = "/etc/webui"
+CONFIG_FILE_NAME = "webuicfg.conf"
+DATABASE_RELATION_NAME = "database"
+DATABASE_NAME = "free5gc"
+
 
 class WebuiOperatorCharm(CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._container_name = self._service_name = "webui"
+        self._container = self.unit.get_container(self._container_name)
+        self._database = DatabaseRequires(
+            self,
+            relation_name=DATABASE_RELATION_NAME,
+            database_name=DATABASE_NAME,
+            extra_user_roles="admin",
+        )
+        self.framework.observe(self.on.webui_pebble_ready, self._on_webui_pebble_ready)
+        self.framework.observe(self.on.database_relation_joined, self._on_webui_pebble_ready)
+        self.framework.observe(self._database.on.database_created, self._on_database_created)
+        self.framework.observe(self._database.on.endpoints_changed, self._on_database_created)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            service_name="webui",
+            ports=[
+                ServicePort(name="urlport-http", port=5000),
+                ServicePort(name="grpc", port=9876),
+            ],
+        )
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created event."""
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        if not self._container.exists(path=BASE_CONFIG_PATH):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        config_file_content = self._render_config_file(
+            database_name=DATABASE_NAME, database_url=event.uris.split(",")[0]
+        )
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+            content=config_file_content
+        ):
+            self._write_config_file(
+                content=config_file_content,
+            )
+        self._on_webui_pebble_ready(event)
+
+    @staticmethod
+    def _render_config_file(database_name: str, database_url: str) -> str:
+        """Renders webui configuration file based on Jinja template.
+
+        Args:
+            database_name: Database Name
+            database_url: Database URL.
+
+        Returns:
+            str: Content of the configuration file.
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template("webuicfg.conf.j2")
+        return template.render(
+            database_name=database_name,
+            database_url=database_url,
+        )
```

Comment:
Charm guidelines mention not using `@staticmethod` and just using plain function in the same file. This could be moved out of the class and removing the annotation.

Reply:
Done